pdg/src/util.rs

```diff
@@ -0,0 +1,132 @@
+use itertools::Itertools;
+
+use std::{
+    cmp::min,
+    collections::HashMap,
+    fmt::{self, Debug, Display, Formatter},
+    hash::Hash,
+};
+
+pub struct ShortOption<T>(pub Option<T>);
+
+impl<T: Display> Display for ShortOption<T> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match &self.0 {
+            Some(this) => write!(f, "{}", this),
+            None => write!(f, "_"),
+        }
+    }
+}
+
+impl<T: Debug> Debug for ShortOption<T> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match &self.0 {
+            Some(this) => write!(f, "{:?}", this),
+            None => write!(f, "_"),
+        }
+    }
+}
+
+pub fn maybe_take<T>(iter: impl Iterator<Item = T>, n: Option<usize>) -> impl Iterator<Item = T> {
+    iter.enumerate()
+        .take_while(move |(i, _)| n.map(|n| *i < n).unwrap_or(true))
+        .map(|(_, t)| t)
+}
+
+pub struct Duplicates<T> {
+    duplicates: HashMap<T, Vec<T>>,
+}
+
+impl<T: Eq + Hash + Clone> Duplicates<T> {
+    pub fn find(iter: impl IntoIterator<Item = T>) -> Self {
+        let duplicates = iter
+            .into_iter()
+            .into_group_map_by(|e| e.clone())
+            .into_iter()
+            .filter(|(_, duplicates)| duplicates.len() > 1)
+            .collect::<HashMap<_, _>>();
+        Self { duplicates }
+    }
+}
+
+impl<T> Duplicates<T> {
+    pub fn is_empty(&self) -> bool {
+        self.duplicates.is_empty()
+    }
+
+    pub fn len(&self) -> usize {
+        self.duplicates.len()
+    }
+}
+
+impl<T> Duplicates<T> {
+    const DEFAULT_UP_TO_N: usize = 5;
+
+    pub fn fmt_up_to_n(
```

Comment:
are there existing libraries that give you something close to this amount of control? seems like a lot of custom code

Reply:
I don't think so.  I customized the error message quite a bit.  Maybe there's some existing crate but I'm not quite sure how to search for something so specific.  `Itertools` is doing the heavy lifting of the actual grouping/deduplication though.

I probably wouldn't have spent this much time on the deduplication test seeing how simple the fix is now, but I hadn't realized it was that simple yet.  Still, I already wrote the test so might as well keep it, and it might help later if we need to switch to on-the-fly deduplication for memory reasons.